scripts/tests/provisioning/test.expect

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/expect -f
+#
+# Start a NILRT VM and instrument installing and running the provisioning
+# ptests. This script assumes that there is a running HTTP server at the ipv4
+# $vm_host_ip, on port 8990.
+# 
+# Expected arguments:
+#   1) path to the run-VM script
+#   2+) ipk feed paths (optional)
+log_file -noappend provisioning.log
+
+set ipk_feeds [lrange $argv 1 end]
+# The SLIRP networking static ipv4 to use for the VM host (as viewed from the
+# guest). Can be any ip on the 10.0.2.0/24 subnet. 10.0.2.2 is the QEMU
+# default, if none is specified.
+set vm_host_ip "10.0.2.2"
+# Set the login username
+set USER root
+
+# boot the VM with the run-vm script
+spawn [lindex $argv 0] -s -- -device e1000,netdev=n1 -netdev user,id=n1,ipv6=off,net=10.0.2.0/24,host=$vm_host_ip
+
+# getty
+# run-postinsts on a non-accelerated VM might take ~300 seconds, so increase
+# the timeout to provide plenty of time.
+set timeout 900
+expect {
+	"login: " { send "root\r" }
+	timeout {
+		send_user "TIMEOUT while waiting for getty."
+		exit 1
+	}
+}
+
+set timeout 90
+# configure opkg
+expect -re "$USER@.*# $" {
+	send "touch /etc/opkg/testing.conf\r"
+	set feed_index 0
+	foreach feed_uri $ipk_feeds {
+		send -- "echo src test-feed.$feed_index http://$vm_host_ip:8990/$feed_uri >>/etc/opkg/testing.conf\r"
+		set feed_index [expr $feed_index + 1]
+	}
+	send "cat /etc/opkg/testing.conf\r"
+	send "opkg update\r"
+}
+
+# setup ptests 
+expect -re "$USER@.*# $" {
+	send "opkg install init-restore-mode-ptest\r"
+}
+
+# run ptests
+expect -re "$USER@.*# $" {
+	send "ptest-runner -l\r"
+	send "ptest-runner\r"
+}
+
+# shutdown
+expect {
+	-re "$USER@.*# $" { send -- "halt\r" }
+	timeout { exit 1 }
+}
+
+expect EOF
```

Comment:
I'm not sure, but should we have a timeout associated with the run-vm script exiting within a reasonable amount of time? For example, if the target VM gets stuck during shutdown should we consider that an error?

This test will probably become a reference for future tests, so if we think that would be useful maybe we want to illustrate it here? If not, feel free to drop this suggestion.

Reply:
The shutdown command will follow the default timeout for the expect script, which is set at 90 seconds. If it takes longer than that, I will print an `TIMEOUT` message and the entire expect process (and VM) will exit with code 1. That will trigger the `set -e` in the `test_provisioning` script and propagate the error to the subcomponent. ie. A hang during shutdown should cause the test to fail sanely.

Reply:
I see. I think I was thrown off by the fact that the expect call just above this had an explicit timeout case. Does that explicit timeout case do anything different than if you had just left it off and relied on the default timeout behavior?

Reply:
See my latest patch-rev.

Reply:
If I understand the behavior properly, I think the **expect eof** still needs a timeout clause if we want to return an error code and print a message.